### PR TITLE
[SPARK-43482][SS] Expand QueryTerminatedEvent to contain error class if it exists in exception

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This adds a new field in QueryTerminatedEvent about "error class" if the information exists in the exception. Note that it doesn't retrieve the information from StreamingQueryException, because it is marked as it's own error class and we want to know the causing exception.

### Why are the changes needed?

Custom streaming query listeners can get the "categorized" error as cause of the query termination, which can be used to determine trends of the errors, aggregation, etc.

### Does this PR introduce _any_ user-facing change?

Yes, users having customer streaming query listener will have additional information in termination event.

### How was this patch tested?

Modified UTs.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @viirya @HyukjinKwon Please take a look. Thanks!

File: python/pyspark/sql/streaming/listener.py [hunk 1]
```diff
@@ -295,6 +299,17 @@ def exception(self) -> Optional[str]:
...
         return self._exception
 
+    @property
+    def errorClassOnException(self) -> Optional[str]:
+        """
+        The error class from the exception if the query was terminated
+        with an exception which is a part of error class framework.
+        If the query was terminated without an exception, or the
+        exception is not a part of error class framework, it will be
+        `None`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/listener.py:309]:
```suggestion
        `None`.

        .. versionadded:: 3.5.0
```

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala [hunk 2]
```diff
@@ -154,11 +154,22 @@ object StreamingQueryListener {
    * @param runId A query id that is unique for every start/restart. See `StreamingQuery.runId()`.
    * @param exception The exception message of the query if the query was terminated
    *                  with an exception. Otherwise, it will be `None`.
+   * @param errorClassOnException The error class from the exception if the query was terminated
+   *                              with an exception which is a part of error class framework.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:158]: Is error class framework a public API that end-users understand or can find document?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:158]: SparkException and SparkThrowable are all public API where we expose to API doc. It's just that we lack about documentation for these classes. That's something we have to fix; we could probably refer to SparkThrowable.getErrorClass here, but it's not documented so wouldn't help unfortunately.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:158]: I think it's fine .. it's sort of documented ..e.g., https://github.com/apache/spark/blob/master/docs/sql-error-conditions-connect-error-class.md

**Review by viirya — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:158]: https://spark.apache.org/docs/latest/sql-error-conditions.html

Above is the landing page on website. It's in SQL guide doc.

**Pull request closed without merging**